### ISSUE TITLE
Add ignoreAtRules option to block-*-brace-* rules

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#561]
- [x] `block-closing-brace-newline-after`
- [x] `block-opening-brace-space-before`

The `ignore` options could accept at-rule names, allowing for SCSS users, for example, to ignore to use `} @else {` syntax with `ignore: ["else"]` on `block-opening-brace-newline-before`.

**Comment by davidtheclark** [comment]
Is there any use-case for this other than `@else`? If not, maybe we should just make `ignore: ["at-else"]`?

**Comment by jeddy3** [comment]
> If not, maybe we should just make ignore: ["at-else"]?

Sgtm. Lets keep it explicit. We can add more choices to the ignore option if other use cases do pop up.

**Comment by evilebottnawi** [comment]
@jeddy3 i scss we have else and else if also. We need options to ignore both. Or may be ignore all at-rule are not included in the official CSS specification. What do you think?

**Comment by davidtheclark** [comment]
> in scss we have else and else if also (for too :)). We need options to ignore both.

If we have `at-else` it would ignore `@else if`, also. It would also ignore `@else anything else you want`, because anything after the first word of the at-rule is a parameter, and we'd just check that the name (`@name`) is `else`.

> Or may be ignore all at-rule are not included in the official CSS specification.

I think that's overkill unless we can think of other at-rules that would also have a unique relation to brackets.

**Comment by evilebottnawi** [comment]
@danieldiekmeier what about `for` in scss?

**Comment by danieldiekmeier** [comment]
@evilebottnawi I think you meant @davidtheclark

**Comment by evilebottnawi** [comment]
@danieldiekmeier sorry

**Comment by davidtheclark** [comment]
@evilebottnawi : I don't understand. Why is `@for` different from other at-rules, in relation to brackets?

**Comment by evilebottnawi** [comment]
@davidtheclark difference code style, i see
```scss
@for $i from 1 through length($colors) {
   ...
}
```
```scss
@for $i from 1 through length($colors) 
{
   ...
}
```
`each` also and etc. I think brackets on this at-rule must lint in plugins for scss, not in stylelint (he ignore non standard keyword).

**Comment by davidtheclark** [comment]
I don't see a really good reason to use that alternate style *only* for
`@for` rules. Seems to me that if you want your at-rule brackets on a new
line instead of the end of the line, that should apply to other at-rules,
also.

**Comment by evilebottnawi** [comment]
@davidtheclark If I have the rule in plugin that tells me to transfer `for` to a new line, but main configuration tell me leaving on the same line, what behavior should I expect?

**Comment by davidtheclark** [comment]
The standard configuration rule is still going to give you a warning, even
if the plugin rule does not, unless you turn off the standard rule or use
ignore comments.

If that variation with `@for` loops is important to you, you actually might
want to create a plugin that conditionally applies the standard `brace`
rule based on your own unique requirements. This is going to be possible
after we release the change for
https://github.com/stylelint/stylelint/issues/709.

**Comment by evilebottnawi** [comment]
@davidtheclark hm, but i want that standard rule to work for me, because my rule lint only on brackets for `for` (`else`, etc). If we had some kind of mechanism hooks, then maybe they could exclude something, or change the behavior some rules. Maybe it seems like overkill, but it would reduce the number of development plugins and increase extensibility.

**Comment by deepfriedmind** [comment]
Just want to add a +1 and agree that being explicit about it is a good idea.

**Comment by jeddy3** [comment]
> If we had some kind of mechanism hooks, then maybe they could exclude something, or change the behavior some rules.

I think I've *finally* grokked what is being asked here (sorry I'm a tad late to the party).

I think there are four options being proposed:

- use *ignore comments* around at-rules that you wish to exclude.
- use [`stylelint.rules`](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules) to conditionally applies the `brace` rules.
- provide a mechanism, *within the secondary options*, for excluding specified at-rules from the `brace` rules.
- provide a mechanism, *within the primary option*,  for changing the behaviour for specified at-rules from the `brace` rules.

I think, after a fair bit of experimenting, that:

- the first option is available and works right now.
- the second option won't work for this particular use case, and correct me if I am wrong though @davidtheclark, as you won't be able to conditionally apply the `brace` rules to *individual at-rules*.
- the third option is seems very reasonable.
- the forth option isn't really viable, and I'll explain why below.

So, a few more details about the third option...

As @davidtheclark pointed out, we already have a convention for excluding things (`ignore: []`). However, I believe that `ignore` has only been used for _predefined_ sets of  _keywords_ (e.g. [`"stylelint-commands"` and `"between-comments"`](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-empty-line-before/README.md#ignore-stylelint-commands). So we might need to use a different convention for user-defined sets, e.g. `ignoreAtRule: []`.

Using this option, you'd be able to *allow* (but not *enforce*) @evilebottnawi's code style of:

```css
a {
  @if $x = 1 {
    color: green;
  } @else if $y = 2 {
    color: blue;
  } @else {
    color: red;
  }
}

@for $i from 1 to 3 
{
  .b-$i { 
    width: $(i)px; 
  }
}

@media (color) {
  a {
    color: red;
  }
}
```

With:

```json
{ 
  "block-closing-brace-newline-after": [
    "always", { 
      "ignoreAtRules": ["if", "else"] 
    }
  ],
  "block-closing-brace-newline-before": "always",
  "block-opening-brace-newline-after": "always",
  "block-opening-brace-space-before": [
    "always", { 
      "ignoreAtRules": ["for"] 
    }
  ]
}
```

The advantage of this approach is that it won't be a breaking change and it will keep the `brace` rules simple to configure for those users who want the same code style across all rules and at-rules.

And so, onto the forth option...

I tried to work through @evilebottnawi's [original proposal](https://github.com/stylelint/stylelint/issues/550#issuecomment-162950322), but it started to unravel. I came across the following issue in your code style that complicates the matter:

```js
{ 
  "block-closing-brace-newline-after": {
    {
      "type": "rule",
      "option": "always"
    }, {
      "type": "at-rule",
      "name": ["keyframes", "media", "for", "else" ...] 
      "option": "always"
    }
  },
  "block-closing-brace-space-after": {
    {
      "type": "at-rule",
      "name": ["if", !!!"else if"!!!] 
      "option": "always"
    }
  },
  ...
}
```
`!!!` - `@else if` behaves differently to `@else` in your code style (i.e. it, like `@if`, has a space after the closing brace), and we can *not* target `@else if` by looking solely at the `atRule.name` as the "else" part comes from `atRule.name` and the "if" part from `atRule.param`.

The object configuration also became unwieldy as I tried to work it though.  So, given:

1. The shortfall I highlighted above.
2. The unwieldiness of the configuration.
3. The complexity of implementation.
4. The suspicion that most users are using a consistent format for all of the their blocks, including for custom at-rules (much like in the [`precss` docs](https://github.com/jonathantneal/precss))

I'd advocate option two (if using *ignore comments* is deemed to be not good enough).

What do you guys think?

@evilebottnawi If we do go with option two, would you like to contribute it to core? You could start with your own use-case first, which I believe means only adding the `"ignoreAtRules: []"` option to the following rules:

- `block-closing-brace-newline-after`
- `block-opening-brace-space-before`

**Comment by davidtheclark** [comment]
> the second option won't work for this particular use case, and correct me if I am wrong though @davidtheclark, as you won't be able to conditionally apply the brace rules to individual at-rules.

Nope, the second option ("use stylelint.rules to conditionally applies the brace rules.") would work. But would be a little involved, as you'd have to write your own plugin.

I am fine with your proposal, @jeddy3, to have a `ignoreAtRules` array. It could accept regexes, too, according to the now-established convention of strings starting and ending with `/`.

**Comment by evilebottnawi** [comment]
@jeddy3 i think `"ignoreAtRules: []"` best solution for this issue. All non-standard properties, I believe should be checked on plugins rules (example for scss). I agree that the detailed configuration of each individual at-rule are superfluous. Perhaps if someone will need configuration of each individual at-rule, the plugins rule is most best solution.

**Comment by jeddy3** [comment]
Good stuff. Lets go with `"ignoreAtRules: []"` (and accept regexes). @evilebottnawi this something you can contribute?

> Nope, the second option ("use stylelint.rules to conditionally applies the brace rules.") would work

@davidtheclark As an aside, I'm still struggling this with. I did underestimated the flexibility of `stylelint.rules` though! I thought that you would only be able to conditionally apply rules to the `root`, but now I understand they _can_ be applied to any node. That's so awesome, and I can now see how powerful it can be for the vast majority of the rules :)

However, I wasn't able to make it work for the `brace` rules as the [rule itself walks the at-rules](https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-newline-after/index.js#L38). And so I couldn't find a way of applying a `brace` rule conditionally-based on the `AtRule` name as you'd have to run the rule on the at-rule's parent, which is likely the root node. Have a missed something here?

Either way, we should definitely, going forward, advocate using `stylelint.rules` when a user has some unique requirements :)

**Comment by evilebottnawi** [comment]
@jeddy3 I would be happy, right now there is no time for it. Sorry :(

**Comment by davidtheclark** [comment]
@jeddy3 You're right. It would be big PITA, but possible, I think.
